url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ColliderImpl.java

```diff
@@ -0,0 +1,798 @@
+                this,
+                sessionSharedData.getInputQueueDataBlockCache(),
+                acceptor,
+                sessionSharedData.getJoinMessageMaxSize(),
+                sessionSharedData.getJoinPool(),
+                m_selector,
+                serverSocketChannel );
+
+        IOException ex = null;
+
+        m_lock.lock();
+        try
+        {
+            if (m_stop)
+                ex = new IOException( "Collider stopped." );
+            else if (m_emitters.containsKey(acceptor))
+                ex = new IOException( "Acceptor already registered." );
+            else
+                m_emitters.put( acceptor, acceptorImpl );
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+
+        if (ex == null)
+            acceptorImpl.start();
+        else
+        {
+            try
+            {
+                serverSocketChannel.close();
+            }
+            catch (final IOException ex1)
+            {
+                /* Should never happen */
+                if (s_logger.isLoggable(Level.WARNING))
+                    s_logger.warning( ex1.toString() );
+            }
+            throw ex;
+        }
+    }
+
+    public void removeAcceptor( Acceptor acceptor ) throws InterruptedException
+    {
+        removeEmitter( acceptor );
+    }
+
+    public void addConnector( Connector connector )
+    {
+        SessionSharedData sessionSharedData = getSessionSharedData( connector );
+
+        ConnectorImpl connectorImpl = new ConnectorImpl(
+                this,
+                sessionSharedData.getInputQueueDataBlockCache(),
+                connector,
+                sessionSharedData.getJoinMessageMaxSize(),
+                sessionSharedData.getJoinPool(),
+                m_selector );
+
+        m_lock.lock();
+        try
+        {
+            if (m_stop)
+                throw new RuntimeException( "Collider is stopped." );
+            else if (m_emitters.containsKey(connector))
+                throw new RuntimeException( "Connector already registered." );
+            else
+                m_emitters.put( connector, connectorImpl );
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+
+        connectorImpl.start();
+    }
+
+    public void removeConnector( Connector connector ) throws InterruptedException
+    {
+        removeEmitter( connector );
+    }
+
+    public void addDatagramListener( DatagramListener datagramListener ) throws IOException
+    {
+        addDatagramListener( datagramListener, null );
+    }
+
+    public void addDatagramListener(
+            DatagramListener datagramListener, NetworkInterface networkInterface ) throws IOException
+    {
+        final InetSocketAddress addr = datagramListener.getAddr();
+        final DatagramChannel datagramChannel = DatagramChannel.open( StandardProtocolFamily.INET );
+        final DatagramSocket socket = datagramChannel.socket();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *socket*. The resource is referenced at lines 676, 679, 680, 686, 699, 700, 704, 705, 706, 736. The resource is closed at lines 694, 765. There are other execution paths that do not contain closure statements, for example, when DatagramChannel.bind throws an exception. Ensure *socket* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).